File: c++/network.cpp
// Network implementation
// mkrum

#include "network.h"
#include <iostream>

Network::Network(int length, int width): output(width){
    for(int i = 0; i < length; i++){
        layers.push_back(Layer(width));
    }
}

double Network::feedForward(vector<double> inputs){
    for(int i = 0; i < layers.size(); i++){
        inputs = layers[i].feedForward(inputs);
    }
    out = output.feedForward(inputs);
    return out;
}

void Network::backPropagate(double target){
    double error = -(target - out);
    auto temp = output.backPropagate(error);
    for(int i = layers.size(); i > 0; i--){
        temp = layers[i].backPropagate(temp);
    }
}

File: scala/network.scala
package network
import neuron.Neuron
import layer.Layer

abstract class Network{
  def layers : Array[Layer]
  def outputNeuron : Neuron 
  var output : Double = _;
  def cost(actual : Double, test : Double) : Double = actual
  def costDer(actual: Double, test : Double) : Double = actual
  def feedForward(ins: Array[Double]) = {
    var inputs = ins;  
    layers foreach { x => inputs = x.feedForward(inputs);}
    outputNeuron.feedForward(inputs)
  }
  def backPropagate(target : Double) {
    var error : Array[Double]  = outputNeuron.backPropagate(costDer(output, target));
    layers.reverse.foreach{ x => error = x.backPropagate(error) };
  } 
  def computeError(values : Array[Array[Double]], targets : Array[Double]) = (values, targets).zipped.map{ case (v : Array[Double], t : Double) => cost(feedForward(v), t)}.sum  
} 
